### PR TITLE
[SW-2677] Make unwrapMojoModel() Independent on Spark Runtime

### PR DESCRIPTION
**Pull request by mn-mikke**
This change allows users to score with the internal mojo model inside `EasyPredictModelWrapper` without running Spark runtime.

### CODE REVIEW

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala [hunk 1]
```diff
@@ -308,32 +311,64 @@ abstract class H2OMOJOModel
...
     config.setModel(mojo)
-    setEasyPredictModelWrapperConfiguration(config)
+    getEasyPredictModelWrapperConfigurationInitializers().foreach(_(config))
     new EasyPredictModelWrapper(config)
   }
 
   override def copy(extra: ParamMap): H2OMOJOModel = defaultCopy(extra)
+
+  val unserializableField = if (System.getProperty("spark.testing", "false") == "true") {
+    new UnserializableClass()
```
Comment (kanech) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:342]: I'm not a native speaker, but I think a correct antonym is NonSerializable, see e.g. https://docs.oracle.com/javase/8/docs/technotes/guides/serialization/examples/nonserialsuper/index3.html
https://docs.unity3d.com/2017.4/Documentation/ScriptReference/NonSerializable.html

Comment (mn-mikke) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:342]: Thanks for catching that!

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOPredictionMultinomial.scala [hunk 2]
```diff
@@ -22,17 +22,26 @@ import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
...
-    val schema = getMultinomialPredictionSchema()
+
+  private val predictionColType = StringType
+  private val predictionColNullable = true
+
+  def getMultinomialPredictionUDF(
+      schema: StructType,
+      modelUID: String,
+      mojoFileName: String,
+      configInitializers: Seq[(EasyPredictModelWrapper.Config) => EasyPredictModelWrapper.Config])
```
Comment (kanech) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOPredictionMultinomial.scala:41]: What about introducing a type alias for: (EasyPredictModelWrapper.Config) => EasyPredictModelWrapper.Config

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OAlgorithmMOJOModel.scala [hunk 3]
```diff
@@ -64,19 +64,19 @@ class H2OAlgorithmMOJOModel(override val uid: String)
...
+  private[sparkling] override def getEasyPredictModelWrapperConfigurationInitializers()
+      : Seq[(EasyPredictModelWrapper.Config) => EasyPredictModelWrapper.Config] = {
+    val superInitializers = super.getEasyPredictModelWrapperConfigurationInitializers()
+    val mojoModel = unwrapMojoModel()
+    val enableContributions = this.getWithContributions() && canGenerateContributions(mojoModel)
+    val enableLeafNodeAssignments = this.getWithLeafNodeAssignments() && canGenerateLeafNodeAssignments(mojoModel)
+    val enableStagedProbabilities = this.getWithStageResults() && canGenerateStageResults(mojoModel)
+
+    superInitializers ++
+      (((config: EasyPredictModelWrapper.Config) => config.setEnableContributions(enableContributions)) ::
```
Comment (krasinski) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OAlgorithmMOJOModel.scala:76]: might be a matter of taste but one way to make it a little shorter could be explicitly stating the type beforehand:
```
    val configInitializers : Seq[EasyPredictModelWrapper.Config => EasyPredictModelWrapper.Config] =
      Seq(_.setEnableContributions(enableContributions),
      _.setEnableLeafAssignment(enableLeafNodeAssignments),
      _.setEnableStagedProbabilities(enableStagedProbabilities))
```

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala [hunk 4]
```diff
@@ -308,32 +311,64 @@ abstract class H2OMOJOModel
...
+    val mojo = unwrapMojoModel()
     config.setModel(mojo)
-    setEasyPredictModelWrapperConfiguration(config)
+    getEasyPredictModelWrapperConfigurationInitializers().foreach(_(config))
     new EasyPredictModelWrapper(config)
   }
 
   override def copy(extra: ParamMap): H2OMOJOModel = defaultCopy(extra)
+
+  val nonSerializableField = if (System.getProperty("spark.testing", "false") == "true") {
```
Comment (krasinski) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:341]: another minot thing - also `.toBoolean` could be used

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala [hunk 5]
```diff
@@ -308,32 +311,64 @@ abstract class H2OMOJOModel
...
+    new NonSerializableClass()
+  } else {
+    null
+  }
+}
+
+class NonSerializableClass extends Serializable {
```
Comment (krasinski) [inline scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:348]: looks a bit confusing, is that piece "test only"?

**Review by krasinski — APPROVED** [review]
LGTM 👍

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala (hunk 5 above)
Comment (mn-mikke) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:348]: Yes, it's only for testing purposes. I want to make sure that H2OMOJOModel is not serialized during scoring and thus it doesn't play role of a closure for scoring udfs.

Comment (mn-mikke) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:348]: I'm open any suggestions for improvemnts.

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala (hunk 4 above)
Comment (mn-mikke) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOModel.scala:341]: will use  thanks!

File: scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOPredictionMultinomial.scala (hunk 2 above)
Comment (mn-mikke) [reply scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOPredictionMultinomial.scala:41]: Thanks for the suggestion, will do!

**Review by kanech — APPROVED** [review]
LGTM

**Pull request merged**